### PR TITLE
Initial hashing service

### PR DESCRIPTION
**Pull request by Pat-ji**
Be sure to call PokemonGo#setHashKey(KEY) directly after creating your
instance.

*MapObjects might need an update to work
*Currently uses the HTTP endpoint, change BosslandHash#BASE if you wish to use a different endpoint
*Might need better failed request handling
*Feel free to improve upon this

### CODE REVIEW

File: library/src/main/java/com/pokegoapi/util/BosslandHash.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        public String json() {
+            Base64.Encoder encoder = Base64.getEncoder();
+            String auth = encoder.encodeToString(authTicket);
+            String session = encoder.encodeToString(sessionData);
+            String request = "";
+            for (byte[] array : requests)
+                request += "\"" + encoder.encodeToString(array) + "\",";
+            request = request.substring(0, request.length() - 1);
+
+            return "{"
```
Comment (PokeAlert) [inline library/src/main/java/com/pokegoapi/util/BosslandHash.java:64]: Why are you not using the already imported Moshi

File: library/src/main/java/com/pokegoapi/util/BosslandHash.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            case 200: // OK
+                String result = response.body().string();
+                response.close();
+                return new HashResponseContent(result);
+            case 400: // Bad request
+                throw new IllegalStateException("Bad request to hashing server, request=" + content.json());
+            case 401: // Unauthorized
+                throw new IllegalStateException("Unauthorized to request hash values");
+            case 429: // Limit exceeded
+                throw new IllegalStateException("Exceeded hashing limit");
```
Comment (PokeAlert) [inline library/src/main/java/com/pokegoapi/util/BosslandHash.java:119]: Not retrying? request lost?

File: library/src/main/java/com/pokegoapi/util/Crypto.java [hunk 3]
```diff
@@ -38,12 +38,8 @@
 
 	private static byte makeIntegrityByte(Rand rand) {
 		rand.state = (0x41C64E6D * rand.state) + 0x3039;
-		long shiftedRand = rand.state >> 16;
-		byte lastbyte = Long.valueOf(shiftedRand).byteValue();
-
-		byte v74 = (byte) ((lastbyte ^ 0x0C) & lastbyte);
-		byte v75 = (byte) (((~v74 & 0x67) | (v74 & 0x98)) ^ 0x6F | (v74 & 8));
-		return v75;
+		byte b = Long.valueOf(rand.state >> 16).byteValue();
```
Comment (PokeAlert) [inline library/src/main/java/com/pokegoapi/util/Crypto.java:41]: I did a test I still get response by removing this
only returning return (byte) (rand.state.byteValue() & 0xE3 | 0x10);
is enough. Can someone confirm?

File: library/src/main/java/com/pokegoapi/util/BosslandHash.java (hunk 1 above)
Comment (Pat-ji) [reply library/src/main/java/com/pokegoapi/util/BosslandHash.java:64]: Didn't have the internet capacity to check out the imported libraries and how they work. I can add it when I get back home

File: library/src/main/java/com/pokegoapi/util/BosslandHash.java (hunk 2 above)
Comment (Pat-ji) [reply library/src/main/java/com/pokegoapi/util/BosslandHash.java:119]: It probably should, this was just a raw convert to get it working. If noone else does it, I'll gladly improve on it when I get back home

**Comment by Gegy** [comment]
#833 has support for both this, and the legacy hasher.

**Pull request closed without merging**